### ISSUE TITLE
复现失败

### ISSUE THREAD

**Issue by Leslie0729** [penghao-wu/visual_jigsaw#8]
你好作者：
     我复现第一个image jigsaw，奖励函数和原文是对齐的，训练数据也是和原文一样的，训练过程中发现reward没有明显的上升趋势，一直保持在0.2左右，只有结构奖励，拼图的奖励几乎没有。除了reward以外，是否还做了其他的修改？？
 {"loss": 0.000909214373677969, "grad_norm": 1.0349620612742858, "learning_rate": 9.203483189051952e-11, "completions/mean_length": 444.328125, "completions/min_length": 273.5, "completions/max_length": 652.0, "completions/clipped_ratio": 0.0, "reward": 0.23385416716337204, "reward_std": 0.01708192378282547, "frac_reward_zero_std": 0.0625, "rewards/JigsawReward/mean": 0.23385417461395264, "rewards/JigsawReward/std": 0.02828053291887045, "kl": 0.09009686950594187, "rollout_correction/training_ppl": 1.79631207883358, "rollout_correction/training_log_ppl": 0.5804733708500862, "rollout_correction/kl": -0.052005983889102936, "rollout_correction/k3_kl": 0.0034405271289870143, "rollout_correction/rollout_ppl": 1.7036967426538467, "rollout_correction/rollout_log_ppl": 0.5282450094819069, "rollout_correction/log_ppl_diff": 0.052228353917598724, "rollout_correction/log_ppl_abs_diff": 0.052228353917598724, "rollout_correction/log_ppl_diff_max": 0.0676775798201561, "rollout_correction/log_ppl_diff_min": 0.03944608196616173, "rollout_correction/ppl_ratio": 1.0536492317914963, "rollout_correction/chi2_token": -0.09094122238457203, "rollout_correction/chi2_seq": -0.09891974925994873, "clip_ratio/low_mean": 0.0, "clip_ratio/low_min": 0.0, "clip_ratio/high_mean": 0.0, "clip_ratio/high_max": 0.0, "clip_ratio/region_mean": 0.0, "epoch": 0.9939540507859734, "step": 2466}, {"loss": 0.0009496198035776615, "grad_norm": 1.2295238883955255, "learning_rate": 6.91289126918404e-11, "completions/mean_length": 455.7734375, "completions/min_length": 312.5, "completions/max_length": 611.5, "completions/clipped_ratio": 0.0, "reward": 0.2395833358168602, "reward_std": 0.018606553319841623, "frac_reward_zero_std": 0.0625, "rewards/JigsawReward/mean": 0.2395833283662796, "rewards/JigsawReward/std": 0.025664037093520164, "kl": 0.09354040212929249, "rollout_correction/training_ppl": 1.8523845672607422, "rollout_correction/training_log_ppl": 0.6091456636786461, "rollout_correction/kl": -0.054257337003946304, "rollout_correction/k3_kl": 0.003613040840718895, "rollout_correction/rollout_ppl": 1.7524289190769196, "rollout_correction/rollout_log_ppl": 0.55463707447052, "rollout_correction/log_ppl_diff": 0.05450858594849706, "rollout_correction/log_ppl_abs_diff": 0.05450858594849706, "rollout_correction/log_ppl_diff_max": 0.07096650451421738, "rollout_correction/log_ppl_diff_min": 0.03969438746571541, "rollout_correction/ppl_ratio": 1.056062325835228, "rollout_correction/chi2_token": -0.09479785338044167, "rollout_correction/chi2_seq": -0.10304789617657661, "clip_ratio/low_mean": 0.0, "clip_ratio/low_min": 0.0, "clip_ratio/high_mean": 0.0, "clip_ratio/high_max": 0.0, "clip_ratio/region_mean": 0.0, "epoch": 0.9947601773478436, "step": 2468}, {"loss": 0.0008836858905851841, "grad_norm": 1.1413152525625956, "learning_rate": 4.94949892400065e-11, "completions/mean_length": 453.3359375, "completions/min_length": 313.0, "completions/max_length": 642.5, "completions/clipped_ratio": 0.0, "reward": 0.24010417610406876, "reward_std": 0.022669488564133644, "frac_reward_zero_std": 0.0, "rewards/JigsawReward/mean": 0.24010416865348816, "rewards/JigsawReward/std": 0.03363868221640587, "kl": 0.08755818381905556, "rollout_correction/training_ppl": 1.7794213891029358, "rollout_correction/training_log_ppl": 0.5710363760590553, "rollout_correction/kl": -0.05138816637918353, "rollout_correction/k3_kl": 0.0034534558944869787, "rollout_correction/rollout_ppl": 1.6884282529354095, "rollout_correction/rollout_log_ppl": 0.5192858539521694, "rollout_correction/log_ppl_diff": 0.051750531885772943, "rollout_correction/log_ppl_abs_diff": 0.051750531885772943, "rollout_correction/log_ppl_diff_max": 0.0671657919883728, "rollout_correction/log_ppl_diff_min": 0.03854966163635254, "rollout_correction/ppl_ratio": 1.0531472116708755, "rollout_correction/chi2_token": -0.089648911729455, "rollout_correction/chi2_seq": -0.0981895700097084, "clip_ratio/low_mean": 0.0, "clip_ratio/low_min": 0.0, "clip_ratio/high_mean": 0.0, "clip_ratio/high_max": 0.0, "clip_ratio/region_mean": 0.0, "epoch": 0.9955663039097138, "step": 2470}, {"loss": 0.000935826450586319, "grad_norm": 1.15062394436564, "learning_rate": 3.313319003717119e-11, "completions/mean_length": 451.875, "completions/min_length": 258.0, "completions/max_length": 631.0, "completions/clipped_ratio": 0.0, "reward": 0.24444444477558136, "reward_std": 0.021878771483898163, "frac_reward_zero_std": 0.0, "rewards/JigsawReward/mean": 0.24444444477558136, "rewards/JigsawReward/std": 0.028952796943485737, "kl": 0.09244657028466463, "rollout_correction/training_ppl": 1.7876782268285751, "rollout_correction/training_log_ppl": 0.5763048902153969, "rollout_correction/kl": -0.05169166810810566, "rollout_correction/k3_kl": 0.0033776046766433865, "rollout_correction/rollout_ppl": 1.696422591805458, "rollout_correction/rollout_log_ppl": 0.5244958251714706, "rollout_correction/log_ppl_diff": 0.051809060387313366, "rollout_correction/log_ppl_abs_diff": 0.051809060387313366, "rollout_correction/log_ppl_diff_max": 0.06518424302339554, "rollout_correction/log_ppl_diff_min": 0.03964153677225113, "rollout_correction/ppl_ratio": 1.0532002300024033, "rollout_correction/chi2_token": -0.09053757786750793, "rollout_correction/chi2_seq": -0.09843414276838303, "clip_ratio/low_mean": 0.0, "clip_ratio/low_min": 0.0, "clip_ratio/high_mean": 0.0, "clip_ratio/high_max": 0.0, "clip_ratio/region_mean": 0.0, "epoch": 0.9963724304715841, "step": 2472}, {"loss": 0.0009411959908902645, "grad_norm": 1.1666333178289965, "learning_rate": 2.0043622169452213e-11, "completions/mean_length": 441.75, "completions/min_length": 273.0, "completions/max_length": 603.5, "completions/clipped_ratio": 0.0, "reward": 0.24270833283662796, "reward_std": 0.020226740278303623, "frac_reward_zero_std": 0.0625, "rewards/JigsawReward/mean": 0.24270834773778915, "rewards/JigsawReward/std": 0.03238314762711525, "kl": 0.09302335232496262, "rollout_correction/training_ppl": 1.8220074623823166, "rollout_correction/training_log_ppl": 0.5947604179382324, "rollout_correction/kl": -0.05330659355968237, "rollout_correction/k3_kl": 0.0036167042853776366, "rollout_correction/rollout_ppl": 1.7257946878671646, "rollout_correction/rollout_log_ppl": 0.5411621183156967, "rollout_correction/log_ppl_diff": 0.053598294500261545, "rollout_correction/log_ppl_abs_diff": 0.053598294500261545, "rollout_correction/log_ppl_diff_max": 0.06810816377401352, "rollout_correction/log_ppl_diff_min": 0.04062163829803467, "rollout_correction/ppl_ratio": 1.0550890862941742, "rollout_correction/chi2_token": -0.09288820251822472, "rollout_correction/chi2_seq": -0.10134239494800568, "clip_ratio/low_mean": 0.0, "clip_ratio/low_min": 0.0, "clip_ratio/high_mean": 0.0, "clip_ratio/high_max": 0.0, "clip_ratio/region_mean": 0.0, "epoch": 0.9971785570334543, "step": 2474}, {"loss": 0.0008792770095169544, "grad_norm": 1.0089467348243557, "learning_rate": 1.0226371306820691e-11, "completions/mean_length": 442.8359375, "completions/min_length": 308.0, "completions/max_length": 647.0, "completions/clipped_ratio": 0.0, "reward": 0.23836806416511536, "reward_std": 0.01646607555449009, "frac_reward_zero_std": 0.0625, "rewards/JigsawReward/mean": 0.23836806416511536, "rewards/JigsawReward/std": 0.02848746906965971, "kl": 0.08745024167001247, "rollout_correction/training_ppl": 1.7863167077302933, "rollout_correction/training_log_ppl": 0.5741444975137711, "rollout_correction/kl": -0.051794255152344704, "rollout_correction/k3_kl": 0.0033314281899947673, "rollout_correction/rollout_ppl": 1.694482460618019, "rollout_correction/rollout_log_ppl": 0.5221247524023056, "rollout_correction/log_ppl_diff": 0.052019762340933084, "rollout_correction/log_ppl_abs_diff": 0.052019762340933084, "rollout_correction/log_ppl_diff_max": 0.06796019151806831, "rollout_correction/log_ppl_diff_min": 0.04072944447398186, "rollout_correction/ppl_ratio": 1.0534277856349945, "rollout_correction/chi2_token": -0.09091186709702015, "rollout_correction/chi2_seq": -0.09873554483056068, "clip_ratio/low_mean": 0.0, "clip_ratio/low_min": 0.0, "clip_ratio/high_mean": 0.0, "clip_ratio/high_max": 0.0, "clip_ratio/region_mean": 0.0, "epoch": 0.9979846835953244, "step": 2476}, {"loss": 0.0009650029242038727, "grad_norm": 1.1637041399607952, "learning_rate": 3.681501702157419e-12, "completions/mean_length": 442.921875, "completions/min_length": 304.5, "completions/max_length": 645.0, "completions/clipped_ratio": 0.0, "reward": 0.23576389998197556, "reward_std": 0.01804919634014368, "frac_reward_zero_std": 0.0625, "rewards/JigsawReward/mean": 0.23576389998197556, "rewards/JigsawReward/std": 0.024601055309176445, "kl": 0.09571702778339386, "rollout_correction/training_ppl": 1.8537984043359756, "rollout_correction/training_log_ppl": 0.6109378263354301, "rollout_correction/kl": -0.053723889868706465, "rollout_correction/k3_kl": 0.0036189244128763676, "rollout_correction/rollout_ppl": 1.7551490664482117, "rollout_correction/rollout_log_ppl": 0.5570635497570038, "rollout_correction/log_ppl_diff": 0.05387428402900696, "rollout_correction/log_ppl_abs_diff": 0.05387428402900696, "rollout_correction/log_ppl_diff_max": 0.07155518233776093, "rollout_correction/log_ppl_diff_min": 0.04203790798783302, "rollout_correction/ppl_ratio": 1.055387243628502, "rollout_correction/chi2_token": -0.09368819929659367, "rollout_correction/chi2_seq": -0.10187712870538235, "clip_ratio/low_mean": 0.0, "clip_ratio/low_min": 0.0, "clip_ratio/high_mean": 0.0, "clip_ratio/high_max": 0.0, "clip_ratio/region_mean": 0.0, "epoch": 0.9987908101571947, "step": 2478}, {"loss": 0.0009000170975923538, "grad_norm": 1.1057166124915454, "learning_rate": 4.0905619086428754e-13, "completions/mean_length": 463.765625, "completions/min_length": 337.0, "completions/max_length": 644.5, "completions/clipped_ratio": 0.0, "reward": 0.2317708432674408, "reward_std": 0.014778628945350647, "frac_reward_zero_std": 0.0, "rewards/JigsawReward/mean": 0.2317708432674408, "rewards/JigsawReward/std": 0.020600808784365654, "kl": 0.08904195949435234, "rollout_correction/training_ppl": 1.817492514848709, "rollout_correction/training_log_ppl": 0.5929963663220406, "rollout_correction/kl": -0.05314790038391948, "rollout_correction/k3_kl": 0.0034950760600622743, "rollout_correction/rollout_ppl": 1.7225303500890732, "rollout_correction/rollout_log_ppl": 0.5398829504847527, "rollout_correction/log_ppl_diff": 0.053113408386707306, "rollout_correction/log_ppl_abs_diff": 0.053113408386707306, "rollout_correction/log_ppl_diff_max": 0.06550217419862747, "rollout_correction/log_ppl_diff_min": 0.04190446063876152, "rollout_correction/ppl_ratio": 1.0545735508203506, "rollout_correction/chi2_token": -0.09302384592592716, "rollout_correction/chi2_seq": -0.10072770901024342, "clip_ratio/low_mean": 0.0, "clip_ratio/low_min": 0.0, "clip_ratio/high_mean": 0.0, "clip_ratio/high_max": 0.0, "clip_ratio/region_mean": 0.0, "epoch": 0.9995969367190649, "step": 2480}], "memory": 58.771484375}

**Comment by Leslie0729** [comment]
你好作者：
        我的实验有2000steps,batchsize=128,rollout=16；奖励函数中结构奖励是0.2，拼图正确奖励是0.8，我的reward curve在下面：
<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/5fbeb2f8-fd9a-4d99-8c12-3c7c40894768" />
         训练后期检查模型的中间思考输出，看起来挺正常，但是最后在boxed{}预测的拼图顺序几乎完全不对，下面是completions的内容：
        "<think>\nTo solve this, I will analyze the images and find common elements or orientation clues that can help me determine the original order of the tiles.  Here are the key observations:\n\n1.  **Tile 1**: Shows a mountainous background with a clear sky.\n2.  **Tile 2**: Shows a portion of the sky and mountain with clouds.\n3.  **Tile 3**: Shows a clear blue sky with no clouds.\n4.  **Tile 4**: Shows a close-up of a person's gloved hand with ski poles.\n5.  **Tile 5**: Shows a close-up of a person with a headband and ski goggles, smiling.\n6.  **Tile 6**: Shows a person wearing ski gear, including a jacket with patches, with mountains in the background.\n7.  **Tile 7**: Shows snow-covered ground with a close-up of ski poles.\n8.  **Tile 8**: Shows a close-up of ski boots on snow.\n9.  **Tile 9**: Shows a partial view of a person's lower body, including ski pants and boots. \n\nBy matching the details:\n- The sky and mountain elements are seen in Tiles 1 and 2. \n- The person is shown in Tiles 5, 6, and possibly 8. \n- Ski gear is visible in Tiles 4, 7, and 8. \n\nThe original order seems to be:\n\n1.  Top left: Mountain and sky (Tile 2)\n2.  Top middle: Clear sky (Tile 3)\n3.  Top right: Mountain and sky (Tile 1)\n\n4.  Middle left: Close-up of ski hand (Tile 4)\n5.  Middle: Close-up of smiling person (Tile 5)\n6.  Middle right: Person with ski gear (Tile 6)\n\n7.  Bottom left: Snow with ski poles (Tile 7)\n8.  Bottom middle: Close-up of boots (Tile 8)\n9.  Bottom right: Partial view of boots and pants (Tile 9)\n\nThus, the correct order is:\n3, 2, 1, 4, 5, 6, 7, 8, 9\n</think>\n\\boxed{3, 2, 1, 5, 6, 4, 7, 8, 9}<|im_end|>"

**Comment by penghao-wu** [comment]
No, we do not make any other modifications. Could you also share your current hyperparameters or the config files?

**Comment by Leslie0729** [comment]
你好作者：
在用你的codebase跑image jigsaw实验时，我发现32轮时间太长了，想问问你第1轮跑完时的reward能上升到多少？能不能把整个32轮的reward curve发一下？

**Comment by penghao-wu** [comment]
@Leslie0729 You can refer to this accuracy curve https://github.com/penghao-wu/visual_jigsaw/issues/4#issuecomment-3473148126
